### DISCUSSION TITLE
Is there a way to replace getQueriesData altogether

### DISCUSSION THREAD

**Discussion by spacepolice10** [TanStack/query#4405] [Q&A] [answered]
Hi there, again.

Sorry to bother, but I can't figure out one thing. TL;DR: 
- I am trying to optimistically update data in one of existing cache instances.
- I use `getQueriesData` to get a tuple with the cache instances I need (one of them has the data I need and I don't know exactly which). 
- Then I update the data in cache with immer and still have a tuple as expected but one of pieces of data is now has properties I want it to have. 
- But when I use `setQueriesData` it is not settings as a tuple but iterated over all the cache instances and the data there replaced with edited tuple which is of course breaks the UI. 

It seems strange to me that you get specific type of data and can't use the same data to update the cache. Is there way around such a behavior so I could update the chosen cache instances altogether? Or is the only way to do so is to iterate over queryKeys, compare them to the corresponding data, generate new sets of data for each key and then use some kind of `for` cycle to update every query properly?

**Comment by TkDodo** [comment, marked as answer]
I'd do:

```
queryClient.setQueriesData(
  ['my', 'fuzzy', 'key'],
  previous => previous?.id === myId ? updateWithImmer(...) : previous
)
```

so the idea is to run `setQueriesData`, which is fuzzy, and it will give you all "old" data for each matching key. Then, you can decide upon that if you want to perform an update or not. You can just return the previous data if it doesn't match what you are looking for.

Note that previous data can be potentially undefined (because you could match queries that are in loading state or have no data because they are disabled). The code I've shown also only works in v4 because only there, you can return undefined from the updater as well.

**Reply by spacepolice10** [reply]
Wow, didn't know that `setQueriesData` returns old data for every cache instance (every key). Might be quite interesting and universal solution. Thank you very much for your help (not only in that case but for all the answers)
